tensorly/decomposition/candecomp_parafac.py

```diff
@@ -247,20 +264,46 @@ def parafac(tensor, rank, n_iter_max=100, init='svd', svd='numpy_svd',\
 
             factors[mode] = factor
 
-        if tol:
-            if sparsity:
-                low_rank_component = kruskal_to_tensor((weights, factors))
-                sparse_component = sparsify_tensor(tensor - low_rank_component, sparsity)
-                
-                unnorml_rec_error = tl.norm(tensor - low_rank_component - sparse_component, 2)
-            else:
-                # ||tensor - rec||^2 = ||tensor||^2 + ||rec||^2 - 2*<tensor, rec>
-                factors_norm = kruskal_norm((weights, factors))
+        # Start line search if requested. Doesn't yet work for fixed modes.
+        if linesearch and iteration % 2 == 0:
+            jump = 1.1
+            total_jump = 0.0
+
+            old_rec_error = error_calc(tensor, weights, factors, sparsity)
+
+            weights_diff = weights - weights_last
+            factors_diff = [factors[ii] - factors_last[ii] for ii in range(tl.ndim(tensor))]
+
+            while True:
+                # Keep stepping while we're successful.
+                new_weights = weights + jump * weights_diff
+                new_factors = [tl.copy(f) for f in factors]
+
+                for ii in modes_list:
+                    new_factors[ii] += factors_diff[ii]
+
+                if mask is not None:
+                    tensor_new = tensor*mask + tl.kruskal_to_tensor((new_weights, new_factors), mask=1-mask)
+                else:
+                    tensor_new = tensor
 
-                # mttkrp and factor for the last mode. This is equivalent to the
-                # inner product <tensor, factorization>
-                iprod = tl.sum(tl.sum(mttkrp*factor, axis=0)*weights)
-                unnorml_rec_error = tl.sqrt(tl.abs(norm_tensor**2 + factors_norm**2 - 2*iprod))
+                new_rec_error = error_calc(tensor_new, new_weights, new_factors, sparsity)
```

Comment:
With your approach you need to compute the error not once, but as many times as you look for a best jump and keep improving. Since computing the cost function for Parafac is as costly as an ALS update, this seems quite expensive? Am I missing something?